src/extensions/windows/Microsoft.DotNet.UpgradeAssistant.Extensions.Windows/WinformsUpdaterSubStep.cs

```diff
@@ -0,0 +1,122 @@
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var currentProject = context.CurrentProject.Required();
+
+            try
+            {
+                var updaterResult = (WinformsUpdaterResult)(await _updater.IsApplicableAsync(context, ImmutableArray<IProject>.Empty.Add(currentProject), token).ConfigureAwait(false));
+                return updaterResult.Result
+                    ? new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"Winforms updater \"{_updater.Title}\" needs to be applied", _updater.Risk)
+                    : new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, string.Empty, BuildBreakRisk.None);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception exc)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                Logger.LogError(exc, "Unexpected exception while initializing Winforms updater \"{WinformsUpdater}\"", _updater.Title);
+                return new UpgradeStepInitializeResult(UpgradeStepStatus.Failed, $"Unexpected exception while initializing Winforms updater \"{_updater.Title}\": {exc}", Risk);
+            }
+        }
+
+        protected override async Task<UpgradeStepApplyResult> ApplyImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var currentProject = context.CurrentProject.Required();
+
+            try
+            {
+                var updaterResult = (WinformsUpdaterResult)(await _updater.ApplyAsync(context, ImmutableArray<IProject>.Empty.Add(currentProject), token).ConfigureAwait(false));
+                if (updaterResult.Result)
+                {
+                    return new UpgradeStepApplyResult(UpgradeStepStatus.Complete, string.Empty);
+                }
+                else
+                {
+                    return new UpgradeStepApplyResult(UpgradeStepStatus.Failed, $"Failed to apply Winforms updater \"{_updater.Title}\"");
+                }
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception exc)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                Logger.LogError(exc, "Unexpected exception while applying Winforms updater \"{WinformsUpdater}\"", _updater.Title);
+                return new UpgradeStepApplyResult(UpgradeStepStatus.Failed, $"Unexpected exception while applying Winforms updater \"{_updater.Title}\": {exc}");
+            }
+        }
+
+        public override async Task<bool> ApplyAsync(IUpgradeContext context, CancellationToken token)
+        {
+            var result = await base.ApplyAsync(context, token).ConfigureAwait(false);
+
+            // Normally, the upgrader will apply steps one at a time
```

Comment:
Is this true? Shouldn't the application just identify that nothing needs to be done and move on to the children?

Reply:
Its start with the children, and when all children are complete the expectation in this case is for the parent step to go green and initialize the next parent step. But without the ApplyAsync call on the parent, it makes the user apply the parent step again.